### PR TITLE
Allow publishers to own referral codes, do not display publisher#promo_stats_2018q1

### PR DESCRIPTION
**Pull request by nvonpentz**
Part two of #1337.  Blocked until https://github.com/brave-intl/publishers/pull/1338 is in production for a couple days to propagate stats to PromoRegistration#stats.

This makes promo registrations polymorphic.  So a channel promo_registration now belongs to one channel and one publisher. A publisher has many promo registrations.

Now we can do things like `publisher.promo_registrations` to get all the promo registrations associated with the publisher. This allows us to easily sum the stats for all promo registrations belonging to a publisher and display them to the publisher.

Part 3 of #1337 will include removing `promo_stats_2018q1q1` from the publishers table and other clean up.

Submitter Checklist:

- [ ] Submitted a [ticket](https://github.com/brave-intl/publishers/issues) for my issue if one did not already exist.
- [ ] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- [ ] Tagged reviewers.
- [ ] Integrated piwik/matomo (for code that adds new buttons).
- [ ] Addressed or ignored all brakeman warnings

Test Plan:

Reviewer Checklist:

Tests
- [ ] Adequate test coverage exists to prevent regressions

Security:
- [ ] No raw SQL -- Always prefer ActiveRecord query helpers ([more info on StackOverflow](https://stackoverflow.com/questions/41410752/rails-5-sql-injection#41452695))
- [ ] XSS is mitigated -- Avoid `html_safe` and `raw`; escape untrusted content from users and 3rd party APIs ([Rails XSS guide](https://brakemanpro.com/2017/09/08/cross-site-scripting-in-rails) and [OWASP XSS Prevention Cheat Sheet](https://www.owasp.org/index.php/XSS_(Cross_Site_Scripting)_Prevention_Cheat_Sheet))

### CODE REVIEW

File: app/models/promo_registration.rb [hunk 1]
```diff
@@ -4,26 +4,29 @@ class PromoRegistration < ApplicationRecord
...
   MONTHLY = "monthly"
   RUNNING_TOTAL = "running_total"
   
   belongs_to :channel, validate: true, autosave: true
   belongs_to :promo_campaign
+  belongs_to :publisher, dependent: :destroy
 
-  validates :channel_id, presence: true, if: -> { kind == CHANNEL}
+  validates :channel_id, presence: true, if: -> { kind == CHANNEL }
+  validates :publisher_id, presence: true, if: -> { kind != UNATTACHED }
```
Comment (corymcdonald) [inline app/models/promo_registration.rb]: Might read a little better if we used `unless`
```suggestion
  validates :publisher_id, presence: true, unless: -> { kind == UNATTACHED }
```

File: app/services/promo_registrar.rb [hunk 2]
```diff
@@ -15,7 +15,11 @@ def perform
         referral_code = result[:referral_code]
         should_update_promo_server = result[:should_update_promo_server]
         if referral_code.present?
-          promo_registration = PromoRegistration.new(channel_id: channel.id, promo_id: @promo_id, kind: "channel", referral_code: referral_code)
+          promo_registration = PromoRegistration.new(channel_id: channel.id,
+                                                     promo_id: @promo_id,
+                                                     kind: "channel",
```
Comment (corymcdonald) [inline app/services/promo_registrar.rb]: Can we use the constant here?
```suggestion
                                                     kind: PromoRegistration::CHANNEL,
```

Comment (nvonpentz) [reply app/services/promo_registrar.rb]: gonna try this commit suggestion feature

**Comment by nvonpentz** [comment]
So we update the old stats here when a publisher logs in:

https://github.com/brave-intl/publishers/blob/2f79dd5651d8f989c420b5c9e74dad0d40bee527/app/controllers/publishers_controller.rb#L367-L369

But since this launches a job, the dashboard won't update until they refresh, so I'm not really sure how effective this is.  (This was previously synchronous and thus guaranteed the stats would be refreshed before the dashboard loads, but we made it async for performance reasons).

I'll need to either remove these lines entirely or replace them with logic that does the same thing with the new system.  What should we do?  I'm tending toward just removing these all together and relying on the Sync::ChannelPromoRegistrationsStatsJob to update the stats in the background.  We could up the frequency of Sync::ChannelPromoRegistrationsStatsJob in sidekiq since it's really only one request per job (where as the previous system was 1 request per publisher within the job).

**Pull request merged**